rpcs3/Emu/NP/rpcn_client.cpp

```diff
@@ -2448,17 +2324,67 @@ namespace rpcn
 
 			if (auto u = friend_infos.friends.find(username); u != friend_infos.friends.end())
 			{
-				if (timestamp < u->second.second)
+				if (timestamp < u->second.timestamp)
 				{
 					break;
 				}
 
-				u->second.second = timestamp;
-				u->second.first  = status;
-				call_callbacks(ntype, username, status);
+				// clear presence data on online/offline
+				u->second = friend_online_data(online, timestamp);
+
+				{
+					std::lock_guard lock(mutex_presence_updates);
+					presence_updates.insert_or_assign(username, u->second);
+				}
+
+				call_callbacks(ntype, username, online);
 			}
 			break;
 		}
+		case NotificationType::FriendPresenceChanged:
+		{
+			auto truncate_string = [](std::string& str, usz max_size)
```

Comment:
considering that this function exists twice, you should probably move this to fmt::truncate.
Also, the `if (>=) resize(size-1)` logic can simply be written as `if (>) resize(size)`